Phone Input: Improvements

umurkontaci opened the pull request:
This PR adds a bunch of improvements to Phone Input and irons out a couple of bugs.

1. Fixes the sanitization issue where the input had <3 chars.
2. Adds alias support for countries that can be called by multiple short codes (e.g. gb==uk)
3. Improves the way to set and update the selected country within the input. We can supply the initial country code as a prop but we need to call a function to make updates. This allows us to let the component to handle its own state, but allow external changes. 
4. Refactored the code for more clarity.

/cc: @klimeryk @aidvu @dzver

aidvu commented on client/components/phone-input/index.jsx:
```diff
@@ -16,21 +18,22 @@ import { countries } from './data';
 	getInitialState() {
-		const selectedCountry = countries[ this.props.selectedCountryCode ];
-		const formattedNumber = formatNumber( this.props.initialValue, selectedCountry );
+		const selectedCountry = countries[ this.props.initialCountryCode ];
```
This is a good idea if no number is passed. Otherwise, I'd say we should try to get the country from the number passed. Found this by reviewing #10313.

umurkontaci replied:
We are also allowing the user to enter the phone number without the country code (in the national format), so there must be a country attached to the phone number at all times.

aidvu replied:
The problem here is:
I have a stored value: +382 67 112233
When I open the form, it turns it into: +1 38267112233

Maybe I didn't use it properly. The idea was, if stored, try to parse country. If cannot, use default. But maybe I'm missing something?

umurkontaci replied:
oh, that makes it a bug :)

umurkontaci replied:
@aidvu Do you have steps to repro?

aidvu replied:
Try to edit contact info for an already registered domain. Or update the contact info to the phone number I provided up there.

What I did is replaced the input field on the Edit Contact Info form with your `PhoneInput` and provided `countriesList` as it's already available there.

umurkontaci commented:
@klimeryk @aidvu @gwwar What do you think about my approach about using methods to get/set selected country instead of `onChange` and `value` props? I think this makes it a little bit easy to use if the parent component doesn't care about the selected country and reduces the number of things that it needs to keep track of. But on the other hand I'm not entirely sure this isn't an anti pattern.

spen commented on client/components/phone-input/index.jsx:
```diff
@@ -39,6 +42,12 @@ const PhoneInput = React.createClass( {
 	},
 
+	componentWillReceiveProps( nextProps ) {
+		if ( this.props.value !== nextProps.value && nextProps.value ) {
```
This looks like it could be a typo?
Is this intentional or should this be `nextProps && nextProps.value` ?

umurkontaci replied:
This checks whether `nextProps.value` is different than `this.props.value` and `nextProps.value` is truthy. 

afaik `nextProps` will always be an object so there's no need to check for `nextProps`.

spen replied:
Ahh of course, silly me! Sorry it was late ;)

umurkontaci replied:
No worries!

spen commented on client/components/phone-input/index.jsx:
```diff
@@ -47,24 +56,33 @@ const PhoneInput = React.createClass( {
 	},
 
+	getFormattedNumberAndCountry( value ) {
+		let formattedNumber = value,
```
The assignment of `formattedNumber = value` seems a little confusing here.
It seems that the only time formattedNumber remains `= value` is when `value` is falsey (it gets reassigned to `formatNumber(...)` otherwise).

Might something like this be clearer?

```es6
let formattedNumber;
let selectedCountry = this.state.selectedCountry;

if ( value ) {
    ...
    formattedNumber = formatNumber( value, selectedCountry );
}
```

Or do we need to pass the falsey value from `value`?

umurkontaci replied:
Good catch, cleared this up.

umurkontaci commented:
I've updated this to get rid of country methods. It's now using onChange and countryCode props to handle country changes. 

I have also fixed the biggest blocker that prevented us from using this widely: cursor jumps. Cursor is now much more stable and you shouldn't experience any jumps while editing characters in between.

klimeryk commented on client/components/phone-input/index.jsx:
```diff
@@ -15,126 +17,125 @@ import { countries } from './data';
+		const nonDigitCountOld = currentFormat
+			.substring( 0, currentCursorPoint )
+			.split( '' )
+			.map( c => isNaN( parseInt( c, 10 ) ) )
```
Could you use a more descriptive name here (and below)? I'm honestly not sure what `c` is supposed to represent here - it's coming from a `currentFormat` variable 😕 Cursor point? Character? :)

umurkontaci commented:
This was part of #7453
